File: code/337.打家劫舍-iii.cpp
/*
 * @lc app=leetcode.cn id=337 lang=cpp
 *
 * [337] 打家劫舍 III
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    unordered_map<TreeNode*, int> sums;
public:
    int rob(TreeNode* root) {
        return tryRob(root);
    }

    int tryRob(TreeNode *node)
    {
        if(!node)
            return 0;
        if(sums.count(node))
            return sums[node];
        
        // 偷当前节点
        int res1 = 0;
        if(node->left)
            res1 += tryRob(node->left->left) + tryRob(node->left->right);
        if(node->right)
            res1 += tryRob(node->right->left) + tryRob(node->right->right);
        res1 += node->val;

        // 不偷当前节点
        int res2 = tryRob(node->left) + tryRob(node->right);
        sums[node] = max(res1, res2);
        return sums[node];
    }
};

File: code/337.打家劫舍-iii.go
/*
 * @lc app=leetcode.cn id=337 lang=golang
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rob(root *TreeNode) int {
	g, f := make(map[*TreeNode]int), make(map[*TreeNode]int)
	dfs(root, g, f)
	return max(g[root], f[root])
}

func dfs(node *TreeNode, g, f map[*TreeNode]int) {
	if node == nil {
		return
	}
	dfs(node.Left, g, f)
	dfs(node.Right, g, f)
	f[node] = node.Val + g[node.Left] + g[node.Right]
	g[node] = max(f[node.Left], g[node.Left]) + max(f[node.Right], g[node.Right])
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
// @lc code=end

